[BUG] CMake incorrectly forces C++14

### Bug summary

I was unable to build DeePMD with PyTorch from source because the code requires C++17 but CMake forces C++14:

https://github.com/deepmodeling/deepmd-kit/blob/e5eac4a68a294838d9933cf7aa6fa2dd8232227d/source/CMakeLists.txt#L24-L27


### DeePMD-kit Version

3.0.1

### Backend and its version

PyTorch

### How did you download the software?

Built from source

### Input Files, Running Commands, Error Log, etc.

Building DeePMD with PyTorch like this:
```
cmake -DENABLE_PYTORCH=TRUE -DCMAKE_CXX_STANDARD=17 -DCMAKE_CXX_STANDARD_REQUIRED=TRUE ..
make -j deepmd_c
```

Leads to errors like these:

```
/home/ole/git/cp2k/tools/toolchain/build/deepmd-kit-3.0.1/source/api_cc/src/DeepSpinPT.cc: In member function ‘void deepmd::DeepSpinPT::compute(ENERGYVTYPE&, std::vector<VALUETYPE>&, std::vector<VALUETYPE>&, std::vector<VALUETYPE>&, std::vector<VALUETYPE>&, std::vector<VALUETYPE>&, const std::vector<VALUETYPE>&, const std::vector<VALUETYPE>&, const std::vector<int>&, const std::vector<VALUETYPE>&, int, const deepmd::InputNlist&, const int&, const std::vector<VALUETYPE>&, const std::vector<VALUETYPE>&, bool)’:
/home/ole/git/cp2k/tools/toolchain/build/deepmd-kit-3.0.1/source/api_cc/src/DeepSpinPT.cc:148:12: error: ‘is_same_v’ is not a member of ‘std’; did you mean ‘is_same’?
  148 |   if (std::is_same_v<VALUETYPE, float>) {
      |            ^~~~~~~~~
      |            is_same
/home/ole/git/cp2k/tools/toolchain/build/deepmd-kit-3.0.1/source/api_cc/src/DeepSpinPT.cc:148:31: error: expected primary-expression before ‘,’ token
  148 |   if (std::is_same_v<VALUETYPE, float>) {
      |                               ^
/home/ole/git/cp2k/tools/toolchain/build/deepmd-kit-3.0.1/source/api_cc/src/DeepSpinPT.cc:148:33: error: expected primary-expression before ‘float’
  148 |   if (std::is_same_v<VALUETYPE, float>) {
      |                                 ^~~~~
/home/ole/git/cp2k/tools/toolchain/build/deepmd-kit-3.0.1/source/api_cc/src/DeepSpinPT.cc:148:32: error: expected ‘)’ before ‘float’
  148 |   if (std::is_same_v<VALUETYPE, float>) {
      |      ~                         ^~~~~~
      |                                )
```

### Steps to Reproduce

```
cmake -DENABLE_PYTORCH=TRUE -DCMAKE_CXX_STANDARD=17 -DCMAKE_CXX_STANDARD_REQUIRED=TRUE ..
make -j deepmd_c
```

### Further Information, Files, and Links

_No response_

--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -15,7 +15,7 @@ set(DEEPMD_C_ROOT
     ""
     CACHE PATH "Path to imported DeePMD-kit C library")
 
-if (NOT DEFINED CMAKE_CXX_STANDARD)
+if(NOT DEFINED CMAKE_CXX_STANDARD)
   set(CMAKE_CXX_STANDARD 11)
 endif()
 macro(set_if_higher VARIABLE VALUE)

--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -15,7 +15,9 @@ set(DEEPMD_C_ROOT
     ""
     CACHE PATH "Path to imported DeePMD-kit C library")
 
-set(CMAKE_CXX_STANDARD 11)
+if (NOT DEFINED CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 11)
+endif()
 macro(set_if_higher VARIABLE VALUE)
   # ${VARIABLE} is a variable name, not a string
   if(${VARIABLE} LESS "${VALUE}")

--- a/source/api_cc/src/DeepPotPT.cc
+++ b/source/api_cc/src/DeepPotPT.cc
@@ -143,7 +143,7 @@ void DeepPotPT::compute(ENERGYVTYPE& ener,
   int natoms = atype.size();
   auto options = torch::TensorOptions().dtype(torch::kFloat64);
   torch::ScalarType floatType = torch::kFloat64;
-  if (std::is_same_v<VALUETYPE, float>) {
+  if (std::is_same<VALUETYPE, float>::value) {
     options = torch::TensorOptions().dtype(torch::kFloat32);
     floatType = torch::kFloat32;
   }
@@ -341,7 +341,7 @@ void DeepPotPT::compute(ENERGYVTYPE& ener,
   int natoms = atype.size();
   auto options = torch::TensorOptions().dtype(torch::kFloat64);
   torch::ScalarType floatType = torch::kFloat64;
-  if (std::is_same_v<VALUETYPE, float>) {
+  if (std::is_same<VALUETYPE, float>::value) {
     options = torch::TensorOptions().dtype(torch::kFloat32);
     floatType = torch::kFloat32;
   }

--- a/source/api_cc/src/DeepSpinPT.cc
+++ b/source/api_cc/src/DeepSpinPT.cc
@@ -145,7 +145,7 @@ void DeepSpinPT::compute(ENERGYVTYPE& ener,
   int natoms = atype.size();
   auto options = torch::TensorOptions().dtype(torch::kFloat64);
   torch::ScalarType floatType = torch::kFloat64;
-  if (std::is_same_v<VALUETYPE, float>) {
+  if (std::is_same<VALUETYPE, float>::value) {
     options = torch::TensorOptions().dtype(torch::kFloat32);
     floatType = torch::kFloat32;
   }
@@ -365,7 +365,7 @@ void DeepSpinPT::compute(ENERGYVTYPE& ener,
   int natoms = atype.size();
   auto options = torch::TensorOptions().dtype(torch::kFloat64);
   torch::ScalarType floatType = torch::kFloat64;
-  if (std::is_same_v<VALUETYPE, float>) {
+  if (std::is_same<VALUETYPE, float>::value) {
     options = torch::TensorOptions().dtype(torch::kFloat32);
     floatType = torch::kFloat32;
   }

